### PR TITLE
Test DINOv2 Implementation - Bug Fixes

### PR DESCRIPTION
**Pull request by yutong-xiang-97**
#### Configs

- Add `embed_dim` and `patch_size` to `method_args`
- Use `embedding_model.wrapped_model`
- Fix the bug in mask height and width (should be `input_size // patch_size`) and update comments

#### Model

- Disable gradients and freeze the teacher and put it to eval mode
- Add masks when doing forward pass to student with global crops
- Fix ibot_head reference bugs
- Add wrappers to ensure the param groups are named differently for the DINO and iBOT heads
- Fix the bug in `TrainableModules` to fix "some parameters appear in more than one parameter group" errors

#### Transform
- Use one transform for all ViTs since we stick to `patch_size=14` and fix corresponding tests

#### Training
- Fix bugs in learning rate scaling and schedule
- Fix bugs in centering
- Move EMA to `on_train_batch_end`
- Remove unused if-clause in getting the layer-wise learning rate decay
- Use `train_loss` instead of `loss` for logging to prevent from errors raised by `EarlyStopping` callback

#### Unit Tests for DINOv2 #115 

- test_optimizer_args_cls
- test_train_step_impl: check if output size matches & dino_head identical to ibot_head when we don't use separate heads
- test_layerwise_decay_optimizer: check if layerwise lrs are correctly set
- test_resolve_auto__scaling_info: check if param scaling in `resolve_auto` works as expected

## How has it been tested?

- Unit Tests
- Be able to reproduce original DINOv2 results.

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_methods/dinov2/dinov2.py [hunk 1]
```diff
@@ -71,6 +71,14 @@ class DINOv2Args(MethodArgs):
         8  # transform_cls().transform_args_cls().transform_args.local_view.num_views
     )
 
+    # vit
+    embed_dim: int = (
+        384  # default embed_dim for ViT-S/B, can be overridden by the model
```
Comment (stegmuel) [inline src/lightly_train/_methods/dinov2/dinov2.py]: For the ViT-B it is 768, if I remember correctly. But this only concerns the comment.

File: src/lightly_train/_methods/dinov2/dinov2.py [hunk 2]
```diff
@@ -71,6 +71,14 @@ class DINOv2Args(MethodArgs):
...
+    )
+    patch_size: int = (
+        14  # default patch_size for ViT-S/B, can be overridden by the model
```
Comment (stegmuel) [inline src/lightly_train/_methods/dinov2/dinov2.py]: Why "for ViT-S/B"? It is for all models, no? This only concerns the comment.

File: src/lightly_train/_methods/dinov2/dinov2.py [hunk 3]
```diff
@@ -331,13 +340,14 @@ def training_step(self, batch: Batch, batch_idx: int) -> Tensor:
...
 
-        loss = training_step_log.loss
+        train_loss = training_step_log.loss
         dino_global_loss = training_step_log.dino_global_loss
         dino_local_loss = training_step_log.dino_local_loss
         ibot_loss = training_step_log.ibot_loss
         koleo_loss = training_step_log.koleo_loss
 
         log_dict = {
+            "train_loss": train_loss,
```
Comment (stegmuel) [inline src/lightly_train/_methods/dinov2/dinov2.py:400]: Are the weight decay and learning rate also logged somewhere? This is quite useful for debugging.

File: src/lightly_train/_methods/dinov2/dinov2.py [hunk 4]
```diff
@@ -525,10 +535,11 @@ def _forward_teacher(
...
-        patch_tokens = patch_tokens.flatten(2).permute(0, 2, 1)  # [G*B, H*W, C]
+        patch_tokens = tokens["features"]  # [G*B, C, H/p, W/p]
+        patch_tokens = patch_tokens.flatten(2).permute(0, 2, 1)  # [G*B, H/p*W/p, C]
+
         masked_patch_tokens = torch.index_select(
-            patch_tokens.flatten(0, 1),  # [G*B*H*W, C]
+            patch_tokens.flatten(0, 1),  # [G*B*H/p*W/p, C]
             dim=0,
             index=mask_indices_list,
         )  # [M, C]
```
Comment (stegmuel) [inline src/lightly_train/_methods/dinov2/dinov2.py:581]: Does `M=G*B*H/p*W/p * 0.5` ?

Comment (yutong-xiang-97) [reply src/lightly_train/_methods/dinov2/dinov2.py:581]: Nope. The half coded 0.5 only means the upperbound of the % of patches masked. You can check how masks are generated: https://github.com/lightly-ai/lightly-train/blob/fb886a5c26d1c166cd2a6548d305f04f28965a74/src/lightly_train/_methods/dinov2/utils.py#L109

File: src/lightly_train/_methods/dinov2/dinov2.py (hunk 3 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_methods/dinov2/dinov2.py:400]: Layerwise lr decay is logged. I added logging of `weight_decay` in the new commit.
You can check them via the MLflow UI (under `dinov2_lightly_train_v0_test`). I used a `weight_decay_end=0.07` as an example.

File: src/lightly_train/_methods/dinov2/dinov2.py [hunk 5]
```diff
@@ -54,6 +58,43 @@
...
+
+def freeze_eval_module(module: Module) -> None:
+    """Freeze the parameters of a module."""
+    for param in module.parameters():
+        param.requires_grad = False
+    module.eval()
+
+
+class DINOHead(Module):
+    """A wrapper for the DINO projection head."""
```
Comment (guarin) [inline src/lightly_train/_methods/dinov2/dinov2.py:73]: Could you add a comment why the wrappers are needed?

File: src/lightly_train/_methods/dinov2/dinov2.py [hunk 6]
```diff
@@ -425,7 +487,7 @@ def training_step_impl(
         )
         student_cls_tokens_global, student_masked_patch_tokens_global = (
             self._forward_student_global(
-                global_views, mask_indices_list
+                global_views, collated_masks, mask_indices_list
```
Comment (guarin) [inline src/lightly_train/_methods/dinov2/dinov2.py]: Nit: Keyword args

File: src/lightly_train/_methods/dinov2/dinov2.py [hunk 7]
```diff
@@ -679,7 +734,7 @@ def on_before_optimizer_step(self, optimizer: Optimizer, *args: Any) -> None:
         self.student_ibot_head.cancel_last_layer_gradients(self.current_epoch)
 
         # Apply weight decay schedule
-        weight_decay = cosine_schedule(
+        self.weight_decay = cosine_schedule(
```
Comment (guarin) [inline src/lightly_train/_methods/dinov2/dinov2.py]: Nit: We shouldn't save this to `self`. If needed we can log it from here directly. But weight decay should actually already be automatically logged from the Lightning. We have a callback/logger for that.

File: src/lightly_train/_methods/dinov2/utils.py [hunk 8]
```diff
@@ -158,44 +160,33 @@ def get_vit_lr_decay_rate(
...
         chunked_blocks (bool): if the blocks are chunked.
     Returns:
         float: lr decay rate for the given parameter.
     """
 
     layer_id = num_layers + 1
-    if name.startswith("backbone") or force_is_backbone:
-        if (
-            ".pos_embed" in name
-            or ".patch_embed" in name
```
Comment (guarin) [inline src/lightly_train/_methods/dinov2/utils.py:171]: Minor: Mention modifications in copyright notice at top

File: tests/_models/dinov2_vit/test_dinov2_vit.py [hunk 9]
```diff
@@ -40,11 +40,23 @@ def test_forward_features(self) -> None:
         feature_extractor = DINOv2ViTModelWrapper(model=model)
 
         x = torch.rand(1, 3, 224, 224)
+        collated_masks = torch.BoolTensor(1, 14 * 14).random_()
```
Comment (guarin) [inline tests/_models/dinov2_vit/test_dinov2_vit.py]: Nit: `BoolTensor` is deprecated. use `torch.tensor(dtype=torch.bool)` instead.

File: tests/_models/dinov2_vit/test_dinov2_vit.py [hunk 10]
```diff
@@ -40,11 +40,23 @@ def test_forward_features(self) -> None:
...
         cls_token = feature_extractor.forward_features(x)["cls_token"]
         assert features.shape == (1, 384, 14, 14)
         assert cls_token.shape == (1, 384)
 
+        features_with_masks = feature_extractor.forward_features(
+            x, masks=collated_masks
+        )["features"]
+        cls_token_with_masks = feature_extractor.forward_features(
+            x, masks=collated_masks
+        )["cls_token"]
```
Comment (guarin) [inline tests/_models/dinov2_vit/test_dinov2_vit.py]: Nit: Can we use a single `forward_features` call?

File: tests/_methods/dinov2/test_dinov2.py [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+@dataclass
+class ScalingResult:
+    output_dim: int
+    start_teacher_temp: float
+    end_teacher_temp: float
+    warmup_teacher_temp_epochs: int
+    momemntum_start: float
+
+
+dummy_vit_model = partial(DINOv2ViTModelWrapper, vit_tiny__testing(patch_size=2))
```
Comment (guarin) [inline tests/_methods/dinov2/test_dinov2.py]: Nit: A function returning the wrapper would be more descriptive

File: src/lightly_train/_models/dinov2_vit/dinov2_vit_src/models/vision_transformer.py [hunk 12]
```diff
@@ -456,3 +456,19 @@ def vit_giant2(patch_size=16, num_register_tokens=0, **kwargs) -> DinoVisionTran
         **kwargs,
     )
     return model
+
+
+def vit_tiny__testing(
```
Comment (guarin) [inline src/lightly_train/_models/dinov2_vit/dinov2_vit_src/models/vision_transformer.py:461]: Good idea! We should do this for all larger models :)

File: tests/_methods/dinov2/test_dinov2.py [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+            "filename": [f"img_{i}" for i in range(b)],
+        }
+
+        # run DistillationV2
+        dinov2_args = DINOv2Args()
+        if ibot_separate_head:
+            dinov2_args.ibot_separate_head = True
+            # To make sure the heads are different
+            dinov2_args.bottleneck_dim_ibot = dinov2_args.bottleneck_dim // 2
+        dinov2_args.n_local_crops = n_local_crops
```
Comment (guarin) [inline tests/_methods/dinov2/test_dinov2.py]: Nit: Pass arguments with `parametrize` to make test clearer and remove logic from function.

**Review by guarin — APPROVED** [review]

File: tests/_methods/dinov2/test_dinov2.py [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+        "scaling, scaling_result",
+        [
+            (
+                ScalingInfo(dataset_size=IMAGENET_SIZE, epochs=100),
+                ScalingResult(
+                    output_dim=65536,
+                    start_teacher_temp=0.04,
+                    end_teacher_temp=0.07,
+                    warmup_teacher_temp_epochs=30,
+                    momemntum_start=0.996,
```
Comment (guarin) [inline tests/_methods/dinov2/test_dinov2.py]: Nit: Typo, same in other places:
```suggestion
                    momentum_start=0.996,
```

File: src/lightly_train/_methods/dinov2/dinov2.py [hunk 15]
```diff
@@ -331,17 +389,21 @@ def training_step(self, batch: Batch, batch_idx: int) -> Tensor:
             batch, batch_idx
         )
 
-        loss = training_step_log.loss
+        train_loss = training_step_log.loss
```
Comment (guarin) [inline src/lightly_train/_methods/dinov2/dinov2.py:393]: We should refactor this ASAP to not have a separate `training_step` implementation for DINOv2.

File: src/lightly_train/_methods/dinov2/dinov2.py (hunk 5 above)
Comment (JonasWurst) [reply src/lightly_train/_methods/dinov2/dinov2.py:73]: addressed

File: src/lightly_train/_methods/dinov2/dinov2.py (hunk 15 above)
Comment (JonasWurst) [reply src/lightly_train/_methods/dinov2/dinov2.py:393]: We can remove this after the debugging has finished imo

File: src/lightly_train/_methods/dinov2/dinov2.py (hunk 6 above)
Comment (JonasWurst) [reply src/lightly_train/_methods/dinov2/dinov2.py]: done

File: tests/_methods/dinov2/test_dinov2.py (hunk 11 above)
Comment (JonasWurst) [reply tests/_methods/dinov2/test_dinov2.py]: done

File: tests/_methods/dinov2/test_dinov2.py (hunk 13 above)
Comment (JonasWurst) [reply tests/_methods/dinov2/test_dinov2.py]: resolved

File: tests/_methods/dinov2/test_dinov2.py (hunk 14 above)
Comment (JonasWurst) [reply tests/_methods/dinov2/test_dinov2.py]: resolved

File: tests/_models/dinov2_vit/test_dinov2_vit.py (hunk 10 above)
Comment (JonasWurst) [reply tests/_models/dinov2_vit/test_dinov2_vit.py]: resolved

File: tests/_models/dinov2_vit/test_dinov2_vit.py (hunk 9 above)
Comment (JonasWurst) [reply tests/_models/dinov2_vit/test_dinov2_vit.py]: solved with `torch.rand(1, 14 * 14) > 0.5`

File: src/lightly_train/_methods/dinov2/utils.py (hunk 8 above)
Comment (JonasWurst) [reply src/lightly_train/_methods/dinov2/utils.py:171]: done

File: src/lightly_train/_methods/dinov2/dinov2.py (hunk 7 above)
Comment (JonasWurst) [reply src/lightly_train/_methods/dinov2/dinov2.py]: I would keep this for debugging at remove later --> keep comment open

File: src/lightly_train/_methods/dinov2/dinov2.py [hunk 16]
```diff
@@ -586,6 +586,9 @@ def _forward_teacher(
             masked_patch_tokens
         )  # [M, D]
 
+        # In order to be compliant with the original DINOv2 code, we need to unsqueeze
+        masked_patch_tokens_after_ibot = masked_patch_tokens_after_ibot.unsqueeze(0)
```
Comment (stegmuel) [inline src/lightly_train/_methods/dinov2/dinov2.py]: Is that compatible with the sinkhorn too?

**Review by stegmuel — COMMENTED** [review]
Is that compatible with the sinkhorn too?

**Review by guarin — APPROVED** [review]

File: src/lightly_train/_methods/dinov2/dinov2.py (hunk 7 above)
Comment (JonasWurst) [reply src/lightly_train/_methods/dinov2/dinov2.py]: Removed from logging and reverted back to old solution (not attached to self)
